### PR TITLE
feat(SPV-892): local operator for regression tests

### PR DESCRIPTION
**Pull request by jakubmkowalski**
# Pull Request Checklist

- [x] 📖 I created my PR using provided  : [CODE_STANDARDS](https://github.com/bitcoin-sv/spv-wallet/blob/main/.github/CODE_STANDARDS.md)
- [x] 📖 I have read the short Code of Conduct: [CODE_OF_CONDUCT](https://github.com/bitcoin-sv/spv-wallet/blob/main/.github/CODE_OF_CONDUCT.md)
- [x] 🏠 I tested my changes locally.
- [ ] ✅ I have provided tests for my changes.
- [x] 📝 I have used conventional commits.
- [x] 📗 I have updated any related documentation.
- [x] 💾 PR was issued based on the Github or Jira issue.

### CODE REVIEW

File: regression_tests/operator.go [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+		fmt.Println("spv-wallet is not running. Run spv-wallet and try again")
+		return
+	}
+
+	sharedConfig, err := GetSharedConfig(adminXPub)
+	if err != nil {
+		fmt.Println("Error getting shared config:", err)
+		return
+	}
+	config.ClientOneURL = sharedConfig.PaymailDomains[0]
```
Comment (wregulski) [inline regression_tests/operator.go:56]: it looks like a repetitive code a bit, maybe consider moving it to separated method? (creating client/operator one and two)

File: regression_tests/operator.go [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+	}
+
+	if !IsSPVWalletRunning(domainLocalHost) {
+		fmt.Println("spv-wallet is not running. Run spv-wallet and try again")
+		return
+	}
+
+	sharedConfig, err := GetSharedConfig(adminXPub)
+	if err != nil {
+		fmt.Println("Error getting shared config:", err)
```
Comment (wregulski) [inline regression_tests/operator.go:53]: minor but start error with lowercase

File: regression_tests/utils.go [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+	ClientTwoURL         string
+	ClientOneLeaderXPriv string
+	ClientTwoLeaderXPriv string
+}
+
+type WalletResponse struct {
+	Message string `json:"message"`
+}
+
+func SaveConfig(config *Config) error {
```
Comment (wregulski) [inline regression_tests/utils.go:51]: please make those methods and structs internal if possible

File: regression_tests/operator.go (hunk 1 above)
Comment (jakubmkowalski) [reply regression_tests/operator.go]: moved to separate method

File: regression_tests/operator.go (hunk 2 above)
Comment (jakubmkowalski) [reply regression_tests/operator.go]: fixed

File: regression_tests/utils.go (hunk 3 above)
Comment (jakubmkowalski) [reply regression_tests/utils.go]: changed to internal

**Review by wregulski — APPROVED** [review]

File: regression_tests/utils.go [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+
+	createPaymailRes, err := adminClient.AdminCreatePaymail(ctx, user.XPub, user.Paymail, "Test test", "")
+	if err != nil {
+		if err.Error() == "paymail address already exists" {
+			return user, fmt.Errorf("paymail address already exists")
+		}
+		return nil, err
+	}
+
+	fmt.Println(keys.XPriv())
```
Comment (Nazarii-4chain) [inline regression_tests/utils.go:217]: Does this needed ?

Comment (jakubmkowalski) [reply regression_tests/utils.go:217]: Can be useful to easily reuse the "leader" account.

File: regression_tests/README.md [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+## Features
+
+- **Users**: Creates, recreates, or uses existing user.
+- **Coin Transfers**: Handle coin transfers to ensure user have the required balance, possibility of self-transfer or transfer with xpriv and domain.
+- **Test Execution**: Run regression tests using either the Go or JS SPV Wallet client.
+- **Configuration Management**: Load (-l flag) and save configuration settings from a `.env.config` file.
+
+## Prerequisites
+
+- **SPV wallet**: up and running on localhost:3003
```
[the remark this answers is not in the document]
Comment (jakubmkowalski) [reply regression_tests/README.md:27]: changed

File: regression_tests/operator.go [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+
+// handleExistingPaymail handles the case when the paymail already exists.
+func handleExistingPaymail(paymailAlias string, config *regressionTestConfig) (*regressionTestUser, error) {
+	if promptUserAndCheck("Paymail already exists. Would you like to use it (you need to have xpriv)? (y/yes or n/no): ") == 1 {
+		return useExistingPaymail(paymailAlias, config)
+	}
+	if promptUserAndCheck("Would you like to recreate user? (y/yes or n/no):") == 1 {
+		return recreateUser(paymailAlias, config)
+	}
+	return nil, fmt.Errorf("can't work with user when xpriv is unknown")
```
[the remark this answers is not in the document]
Comment (jakubmkowalski) [reply regression_tests/operator.go:133]: changed

File: regression_tests/operator.go [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+		}
+	} else {
+		if err := takeMasterUrlAndXPriv(user); err != nil {
+			return fmt.Errorf("error sending coins: %v", err)
+		}
+	}
+
+	leaderBalance := 0
+	for leaderBalance == 0 {
+		fmt.Print("Waiting for coins")
```
[the remark this answers is not in the document]
Comment (jakubmkowalski) [reply regression_tests/operator.go:180]: changed

[the remark this answers is not in the document]
Comment (jakubmkowalski) [reply regression_tests/operator.go:180]: changed

File: regression_tests/README.md [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+- [Installation](#installation)
+- [Configuration](#configuration)
+- [Usage](#usage)
+- [Operator Flow](#operator-flow)
+- [License](#license)
+
+## Features
+
+- **Users**: Creates, recreates, or uses existing user.
+- **Coin Transfers**: Handle coin transfers to ensure user have the required balance, possibility of self-transfer or transfer with xpriv and domain.
```
Comment (Nazarii-4chain) [inline regression_tests/README.md:21]:
```suggestion
- **Funds Transfers**: Handle funds transfers to ensure user have the required balance, possibility of self-transfer or transfer with xpriv and domain.
```

Comment (jakubmkowalski) [reply regression_tests/README.md]: changed

**Review by Nazarii-4chain — APPROVED** [review]

File: regression_tests/operator.go [hunk 9]
```diff
@@ -0,0 +1,262 @@
...
+	if *loadConfigFlag {
+		config, err = loadConfig()
+		if err != nil {
+			fmt.Println("error loading config:", err)
+			return
+		}
+		user.XPriv = config.ClientOneLeaderXPriv
+	} else {
+		config = &regressionTestConfig{}
+	}
```
Comment (chris-4chain) [inline regression_tests/operator.go:44]: Consider making a utility function for that part of code :)

File: regression_tests/operator.go [hunk 10]
```diff
@@ -0,0 +1,262 @@
...
+	if err != nil {
+		return handleCreateUserError(err, paymailAlias, config)
+	}
+
+	return user, nil
+}
+
+// handleCreateUserError handles the error when creating a user.
+func handleCreateUserError(err error, paymailAlias string, config *regressionTestConfig) (*regressionTestUser, error) {
+	if err.Error() == "paymail already exists" {
```
Comment (chris-4chain) [inline regression_tests/operator.go:119]: Maybe you could define errors once and then check for the specific error using `errors.Is`. 

That approach would prevent from undefined behavior when you change the actual string for some error.

File: regression_tests/operator.go [hunk 11]
```diff
@@ -0,0 +1,262 @@
...
+
+// handleFundsTransfer handles the transfer of funds to a user.
+func handleFundsTransfer(user *regressionTestUser, config *regressionTestConfig) error {
+	response := promptUserAndCheck("Do you have xpriv and master instance URL? (y/yes or n/no): ")
+	if response == 0 {
+		fmt.Printf("Please send %d Sato for full regression tests:\n%s\n", minimalBalance, user.Paymail)
+		isSent := 0
+		for isSent < 1 {
+			isSent = promptUserAndCheck("Did you make the transaction? (y/yes or n/no): ")
+		}
```
Comment (chris-4chain) [inline regression_tests/operator.go:172]: Maybe we should define a way to cancel and/or add a timeout?

File: regression_tests/operator.go [hunk 12]
```diff
@@ -0,0 +1,262 @@
...
+	leaderBalance := 0
+	for leaderBalance == 0 {
+		fmt.Print("Waiting for funds")
+		for i := 0; i < 3; i++ {
+			fmt.Print(".")
+			time.Sleep(1 * time.Second)
+		}
+		leaderBalance = checkBalance(config.ClientOneURL, config.ClientOneLeaderXPriv)
+		fmt.Println()
+	}
```
Comment (chris-4chain) [inline regression_tests/operator.go:188]: Maybe you could print a leader balance when it's nonzero?

File: regression_tests/operator.go (hunk 9 above)
Comment (chris-4chain) [reply regression_tests/operator.go:44]: After DM conversation - we'll maybe do it in the second iteration

File: regression_tests/operator.go (hunk 10 above)
Comment (chris-4chain) [reply regression_tests/operator.go:119]: After DM conversation - we'll maybe do it in the second iteration

File: regression_tests/operator.go (hunk 12 above)
Comment (chris-4chain) [reply regression_tests/operator.go:188]: After DM conversation - we'll maybe do it in the second iteration

**Review by chris-4chain — APPROVED** [review]

File: regression_tests/operator.go (hunk 9 above)
Comment (jakubmkowalski) [reply regression_tests/operator.go]: Changed to unresolved as more changes were needed. 
Utility function made.

File: regression_tests/operator.go (hunk 11 above)
Comment (jakubmkowalski) [reply regression_tests/operator.go]: timeout added

File: regression_tests/operator.go (hunk 10 above)
Comment (jakubmkowalski) [reply regression_tests/operator.go]: Changed to unresolved as more changes were needed. 
For now only predefinition of error was made (errors.Is will be introduced after fixes of errors will be merged).

File: regression_tests/utils.go [hunk 13]
```diff
@@ -0,0 +1,324 @@
...
+			if checkResult != -1 {
+				return checkResult, nil
+			}
+			fmt.Println("Invalid response. Please answer y/yes or n/no.")
+		}
+	}
+}
+
+// checkResponse checks the response and returns an integer indicating the result.
+func checkResponse(response string) int {
```
Comment (chris-4chain) [inline regression_tests/utils.go:182]: Consider some `yes`, `no`, `noChoice` enum

File: regression_tests/operator.go [hunk 14]
```diff
@@ -0,0 +1,301 @@
...
+	var err error
+	if *loadConfigFlag {
+		config, err = getEnvVariables()
+		if err != nil {
+			fmt.Println("error loading config:", err)
+			return nil, err
+		}
+	}
+	return config, nil
+
```
[the remark this answers is not in the document]
Comment (jakubmkowalski) [reply regression_tests/operator.go]: removed

File: regression_tests/operator.go [hunk 15]
```diff
@@ -0,0 +1,301 @@
...
+	if err := runTests(clientType, defaultPath); err != nil {
+		fmt.Println("error running tests:", err)
+	}
+}
+
+// getConfig retrieves the configuration from the environment or from the user.
+func getConfig(loadConfigFlag *bool, config *regressionTestConfig) (*regressionTestConfig, error) {
+	var err error
+	if *loadConfigFlag {
```
[the remark this answers is not in the document]
Comment (jakubmkowalski) [reply regression_tests/operator.go]: changed to standard bool without pointer

File: regression_tests/operator.go [hunk 16]
```diff
@@ -0,0 +1,301 @@
...
+		config, err = getEnvVariables()
+		if err != nil {
+			fmt.Println("error loading config:", err)
+			return nil, err
```
[the remark this answers is not in the document]
Comment (jakubmkowalski) [reply regression_tests/operator.go]: Message removed, all occurrences of `fmt.Println("error...` changed to returns of error to main function, and then those are printed, right before program ends.

File: regression_tests/operator.go [hunk 17]
```diff
@@ -0,0 +1,301 @@
...
+}
+
+// handleUserCreation handles the creation of a user.
+func handleUserCreation(paymailAlias string, config *regressionTestConfig) (*regressionTestUser, error) {
+	if config.ClientOneLeaderXPriv != "" {
+		answer, err := promptUserAndCheck("Would you like to use user from env? (y/yes or n/no): ")
+		if err != nil {
+			return nil, fmt.Errorf("failed to prompt user: %w", err)
+		}
+		if answer == 1 {
```
[the remark this answers is not in the document]
Comment (jakubmkowalski) [reply regression_tests/operator.go]: Constants introduced to these values.

File: regression_tests/operator.go [hunk 18]
```diff
@@ -0,0 +1,301 @@
...
+		fmt.Printf("Please send %d Sato for full regression tests:\n%s\n", minimalBalance, user.Paymail)
+		isSent := -1
+		for isSent < 0 {
+			isSent, err = promptUserAndCheck("Did you make the transaction? (y/yes or n/no): ")
+			if isSent == 0 {
+				fmt.Println("Checking balance")
+			}
+		}
+	} else {
+		if err := takeMasterUrlAndXPriv(user); err != nil {
```
[the remark this answers is not in the document]
Comment (jakubmkowalski) [reply regression_tests/operator.go]: changed

File: regression_tests/operator.go [hunk 19]
```diff
@@ -0,0 +1,301 @@
...
+
+// takeMasterUrlAndXPriv takes the master URL and xpriv for transferring funds.
+func takeMasterUrlAndXPriv(leaderPaymail *regressionTestUser) error {
+	url := getValidURL()
+	xprivMaster := getValidXPriv()
+
+	err := sendFundsWithGoClient(url, xprivMaster, leaderPaymail.Paymail)
+	if err != nil {
+		fmt.Println("error sending funds:", err)
+		return err
```
[the remark this answers is not in the document]
Comment (jakubmkowalski) [reply regression_tests/operator.go]: Now errors are handled and printed in main function.

File: regression_tests/utils.go [hunk 20]
```diff
@@ -0,0 +1,324 @@
...
+
+	for {
+		timeout := time.After(timeoutDuration)
+		input := make(chan string)
+		go func() {
+			fmt.Println(question)
+			response, _ = reader.ReadString('\n')
+			response = strings.TrimSpace(response)
+			input <- response
+		}()
```
[the remark this answers is not in the document]
Comment (jakubmkowalski) [reply regression_tests/utils.go]: `reader.ReadString` as You said will not be stopped. Before, timeout error was passed to main function and then execution of program was ended. Now to make it clear what happens at the timeout os.Exit has been added there.

File: regression_tests/utils.go (hunk 13 above)
Comment (jakubmkowalski) [reply regression_tests/utils.go:185]: Constants introduced to these values.

**Review by chris-4chain — APPROVED** [review]

**Pull request merged**